### PR TITLE
SONARJS-867 SE: Track whether numbers can be positive or negative

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/se/points/UnaryNumericProgramPoint.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  public UnaryNumericProgramPoint(Tree element, SymbolicExecution execution) {
+    this.element = (UnaryExpressionTree) element;
+    this.execution = execution;
+  }
+
+  @Override
+  public ProgramState execute(ProgramState state) {
+    ExpressionStack stack = state.getStack();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/points/UnaryNumericProgramPoint.java]: MINOR Remove this use of "getStack"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/Relation.java [hunk 2]
```diff
@@ -127,6 +138,46 @@ public String operatorString() {
...
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    private static Optional<Boolean> greaterThan(Range<Integer> range1, Range<Integer> range2) {
+      return lessThan(range2, range1);
+    }
+
+    private static Optional<Boolean> not(Optional<Boolean> value) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Relation.java]: MINOR Specify a "Boolean" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ExpressionStack.java#L162](https://github.com/SonarSource/sonar-javascript/blob/bdfbc2d02da3dcfd11e14d82f6759af702a1e41f/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L162): This case's code block is the same as the block for the case on line 118. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ExpressionStack.java#L162](https://github.com/SonarSource/sonar-javascript/blob/d7d3f3a58005b8794bbaff1ed0f29e5b3c703f21/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L162): This case's code block is the same as the block for the case on line 118. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ExpressionStack.java#L162](https://github.com/SonarSource/sonar-javascript/blob/10253ba6c4787de599f2cb70f12b2037c8d94f79/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L162): This case's code block is the same as the block for the case on line 118. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: javascript-frontend/src/main/java/org/sonar/javascript/se/sv/LiteralSymbolicValue.java [hunk 3]
```diff
@@ -69,7 +69,7 @@ private static Constraint getConstraint(LiteralTree literal) {
       result = literal.value().length() > 2 ? Constraint.TRUTHY_STRING_PRIMITIVE : Constraint.EMPTY_STRING_PRIMITIVE;
     }
     if (literal.is(Kind.NUMERIC_LITERAL)) {
-      result = isTruthyNumeric(literal) ? Constraint.TRUTHY_NUMBER_PRIMITIVE : Constraint.ZERO;
+      result = isTruthyNumeric(literal) ? Constraint.POSITIVE_NUMBER_PRIMITIVE : Constraint.ZERO;
```
Comment (inverno) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/sv/LiteralSymbolicValue.java]: Shouldn't we rename this one to `isPositiveNumeric` ?

File: javascript-frontend/src/main/java/org/sonar/javascript/se/points/UnaryNumericProgramPoint.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  private SymbolicValue assignedValue = null;
+
+  public UnaryNumericProgramPoint(Tree element, SymbolicExecution execution) {
+    this.element = (UnaryExpressionTree) element;
+    this.execution = execution;
+  }
+
+  @Override
+  public Optional<ProgramState> execute(ProgramState state) {
+    ExpressionStack stack = state.getStack();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/points/UnaryNumericProgramPoint.java]: MINOR Remove this use of "getStack"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/points/ProgramPoint.java [hunk 5]
```diff
@@ -21,13 +21,14 @@
...
 import org.sonar.javascript.se.ProgramState;
+import org.sonar.javascript.se.SymbolicExecution;
 import org.sonar.plugins.javascript.api.tree.Tree;
 
 public interface ProgramPoint {
 
   Optional<ProgramState> execute(ProgramState state);
 
-  static ProgramPoint create(Tree element) {
+  static ProgramPoint create(Tree element, SymbolicExecution execution) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/points/ProgramPoint.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ExpressionStack.java#L162](https://github.com/SonarSource/sonar-javascript/blob/93e957dbb36c8465a6a2324f8f65bdea956ef14d/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L162): This case's code block is the same as the block for the case on line 118. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: javascript-frontend/src/main/java/org/sonar/javascript/se/points/UnaryNumericProgramPoint.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        this.expressionValue = convertToNumber(operandValue, operandConstraint);
+
+      } else {
+        this.expressionValue = new UnaryMinusSymbolicValue(operandValue);
+
+      }
+    }
+  }
+
+  private SymbolicValue convertToNumber(SymbolicValue operandValue, Constraint operandConstraint) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/points/UnaryNumericProgramPoint.java]: MAJOR Make "convertToNumber" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ExpressionStack.java#L162](https://github.com/SonarSource/sonar-javascript/blob/8b955545df336c0bff097c71c512cc4f65cb2bc1/javascript-frontend/src/main/java/org/sonar/javascript/se/ExpressionStack.java#L162): This case's code block is the same as the block for the case on line 118. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Pull request merged**